### PR TITLE
remove the unsafe os.chdir(), and make sure we clean up the workdir w…

### PR DESCRIPTION
**Pull request by SvenDowideit**
Signed-off-by: Sven Dowideit <sven.dowideit@csiro.au>

# Overview
`_parse_and_execute` can call `process.clean()` so chdir-ing to the mkdtemp() can cause problems when processing multiple requests. Also moved the `process.clean()` to `execute()` to ensure that we actually clean up the workdir we create there - irrespective of exceptions thrown higher up.

# Related Issue / Discussion
* #235, #228

(from my comment in #235:)

I'm hitting this too - and in poking around, it seems to me the root problem is that the code is doing an `os.chdir` at all - see for example, 

* https://modwsgi.readthedocs.io/en/develop/user-guides/application-issues.html#application-working-directory
* https://stackoverflow.com/questions/12081789/pythons-working-directory-when-running-with-wsgi-and-apache

if you create multiple outstanding requests, and then watch the value of both os.curdir and olddir in all of the requests, you'll see that olddir eventually gets set to another process's workdir - which also means that any processing assuming its curdir will start reading and writing to somewhere random.

introduced in
 * https://github.com/geopython/pywps/commit/062aeb12344d4df677e2ad9e56ac151bc4160f12#diff-1ab67f2df437e13a3ace4c4c031e1a68
* https://github.com/geopython/pywps/commit/52dc170242fbad625fdf1339850228c7d4e449c5#diff-1ab67f2df437e13a3ace4c4c031e1a68, then modified in 
* https://github.com/geopython/pywps/commit/f67ab47548cf8ad3d75398d52fc3b7f6c45f890e#diff-1ab67f2df437e13a3ace4c4c031e1a68 and moved to finally in 
* https://github.com/geopython/pywps/commit/1e73102786d51e340bbe7ebd90af62a1cc08173c#diff-1ab67f2df437e13a3ace4c4c031e1a68

# Contribution Agreement

- [x] I'd like to contribute this bugfix to PyWPS. I confirm that my contributions to PyWPS will be compatible with the PyWPS license guidelines at the time of contribution.

### CODE REVIEW

File: pywps/app/Service.py [hunk 1]
```diff
@@ -301,12 +301,10 @@ def execute(self, identifier, wps_request, uuid):
         except KeyError:
             raise InvalidParameterValue("Unknown process '%r'" % identifier, 'Identifier')
 
-        olddir = os.path.abspath(os.curdir)
         try:
-            os.chdir(process.workdir)
             response = self._parse_and_execute(process, wps_request, uuid)
         finally:
-            os.chdir(olddir)
+            process.clean()
```
Comment (elemoine) [inline pywps/app/Service.py]: If the case of an async job isn't this going to remove the `workdir` before the job has actually run?

The [ExecuteResponse object](https://github.com/geopython/pywps/blob/bdd404e7af9cac7e343e3ebf32111a3d43dd8137/pywps/response/execute.py) is currently responsible for calling `clean` on the process. So what's the reason for calling `clean` here?

File: pywps/app/Service.py [hunk 2]
```diff
@@ -381,7 +379,6 @@ def _parse_and_execute(self, process, wps_request, uuid):
                 for proc_outpt in process.outputs:
                     if outpt == proc_outpt.identifier:
                         resp = Response(proc_outpt.data)
-                        resp.call_on_close(process.clean)
```
Comment (elemoine) [inline pywps/app/Service.py]: I agree with this change, but I'd also remove the `call_on_close` function entirely, and I'd put that in a separate commit from the one removing `os.chdir(process.workdir)`.

File: pywps/app/Service.py (hunk 1 above)
Comment (SvenDowideit) [reply pywps/app/Service.py]: oh my! my "reason" for cleaning up here, is that its in the same area of code as the creation - but by adding async - and having an ExecuteResponse cleanup, its redundant.

File: pywps/app/Service.py (hunk 2 above)
Comment (SvenDowideit) [reply pywps/app/Service.py]: interesting - I'll go revisit in a moment

File: pywps/app/Service.py [hunk 3]
```diff
@@ -92,12 +92,7 @@ def execute(self, identifier, wps_request, uuid):
         except KeyError:
             raise InvalidParameterValue("Unknown process '%r'" % identifier, 'Identifier')
 
-        olddir = os.path.abspath(os.curdir)
-        try:
-            os.chdir(process.workdir)
-            response = self._parse_and_execute(process, wps_request, uuid)
-        finally:
-            os.chdir(olddir)
+        response = self._parse_and_execute(process, wps_request, uuid)
```
Comment (elemoine) [inline pywps/app/Service.py]: I get `response = None` above now serves no purpose.

Comment (SvenDowideit) [reply pywps/app/Service.py]: fair!

**Comment by elemoine** [comment]
This PR looks good to me. And it fixes a real issue in PyWPS.

**Comment by elemoine** [comment]
@cehbrecht have you seen the related issues #235 and #228? And have you read https://github.com/geopython/pywps/issues/235#issuecomment-357819074 and the links @SvenDowideit provided there? This is what the mod_wsgi documentation says for example: "An application should also never change the current working directory and then assume that it can then use relative paths.".

**Comment by cehbrecht** [comment]
@elemoine I can understand the issue you would like to fix. But when I apply the current patch all process execution requests will use the same working directory. So, we need an improved version. 

Maybe the PyWPS people have some comments.

**Comment by cehbrecht** [comment]
... some more comments. 

The way PyWPS works currently it is creating a temporary directory (working dir) for each process execution. It changes into this temporary directory ... so in my process I can assume that the current dir is the temp folder I can use for my temp files. When my process has finished (in what ever state) this temp folder gets removed.

As I understand from this patch we shouldn't change into directories (`os.chdir()`). But I still want to have my temp folder for my process execution which pywps handles for me.  So, pywps would still create a temp folder before running the process and removes it after the process has finished. But it wouldn't change the directory. In my process I cannot use the current directory anymore ... I would need to use a provided variable `working_dir` to store my temp files.

Does this make sense?

**Comment by SvenDowideit** [comment]
@cehbrecht This change does not change the creation of a temp dir - that is still created, and made available as the workdir - and yes, this is what users should / must use.

Users should never put files into the current directory - this is true for pretty much all web based toolings - from the ancient CGI, and to this date - ideally, the current dir should be set read only by the web server admin.

The way the code currently works, the current dir for all currently in progress requests is changed whenever there is a new request - which means there are race conditions which could result in some requests returning the wrong data, and most commonly exhibits in the less harmful way as reported in #238 and #228.

so to sum up - yes, pywps does make a temporary workdir, which it cleans up after the request processing has completed - users need to be intentional about where they write data (to `working_dir`), otherwise multiple outstanding requests can cause indeterminate results.

**Comment by elemoine** [comment]
@SvenDowideit's comments make perfect sense to me. And the issues reported in #238 and #228 are real! To me this change should be merged. It is bad to change the web server's current dir.

**Comment by cehbrecht** [comment]
The PyWPS people have the last word on it :) I'm just commenting. 

As a follow-up of this PR we should make this clear in the docs and provide and an example how to use the temp-folder. Before it was quite convenient to use just the current dir ... I need to adapt my processes.

**Comment by elemoine** [comment]
@cehbrecht you are right about that.

**Comment by ldesousa** [comment]
Thank you for this contribution, and apologies for the time it took to merge.
